pippy/auto_parallelization.py

```diff
@@ -0,0 +1,265 @@
+# Adaptation of https://github.com/alpa-projects/alpa/blob/a88992ce3b46024c0a4ee4aa8cb069a62830cec2/alpa/pipeline_parallel/stage_construction.py
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from dataclasses import dataclass
+from typing import List, Tuple
+
+import numpy as np
+import torch.fx
+from enum import Enum
+
+from pippy import pipe_split
+
+try:
+    from numba import njit
+    from numba.typed import List as NumbaList
+except ImportError:
+    def njit(*args, **kwargs):
```

Comment:
decorators that take args have to return a function that [actually does the wrapping](https://stackoverflow.com/a/5929165)